x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authc/jwt/JwtAuthenticationToken.java

```diff
@@ -95,4 +93,42 @@ public void clearCredentials() {
     public String toString() {
         return JwtAuthenticationToken.class.getSimpleName() + "=" + this.principal;
     }
+
+    private String buildTokenPrincipal() {
+        JWTClaimsSet jwtClaimsSet = getJWTClaimsSet();
+        StringBuilder principalBuilder = new StringBuilder();
+        for (String claimName : new TreeSet<>(jwtClaimsSet.getClaims().keySet())) {
```

Comment:
I think should just base64 decode the middle part (the raw claim set) use that as-is.  org.elasticsearch.xpack.security.authc.jwt.JwtUtil#toStringRedactSignature has an example of working with the individual parts of the encoded string. The audit log output will be JSON but should still tokenize fairly well for search-ability and functionally makes sense that JSON is how we identify a JSON web token.  I have minor concerns about using the value as a cache key (size and non-garunteed ordering) but those concerns are present with this implementation too and probably isn't that much of a concern in practice. (ideally we would introduce a cache key which could just be a hash of the principal, but that is probably outside the scope of this PR) (in practice, even though order is not guaranteed, i would assume (mostly) consistent ordering for the same JWT in the same JVM).

In general, we need to be extra careful about exception handling processing for performance sensitive areas. Since this is running on the transport thread and part of auth workflow raising exceptions as part of non-exceptional workflows can be really expensive due to the need to grab the stack trace (even if it is not used). If we continue this approach we should prefer manual parsing and/or instance of checks if the API requires you to know the claim type up front (vs. just a map of key/values) .

Reply:
> In general, we need to be extra careful about exception handling processing for performance sensitive areas. Since this is running on the transport thread and part of auth workflow raising exceptions as part of non-exceptional workflows can be really expensive due to the need to grab the stack trace (even if it is not used). If we continue this approach we should prefer manual parsing and/or instance of checks if the API requires you to know the claim type up front (vs. just a map of key/values) .

I've pushed [2b47c9d](https://github.com/elastic/elasticsearch/pull/101333/commits/2b47c9d443a7ec5d884beea30a57474f7879eac2) which doesn't use exception throwing when parsing the token principal from the JWT token.

Reply:
I have 2 objections to displaying the JWT as a JSON in the `user.name` audit event field:
1) it's long and contains date claims such as `exp`, `nbf`, `iat` which are not useful to identifying the principal (in addition to other claims like that, but that we cannot clearly distinguish)
2) the JWTJSON gets escaped as a string inside the `user.name` JSON field: https://github.com/elastic/elasticsearch/blob/676e28e50b50c5242fdc6cbb860771bb472e3a68/x-pack/plugin/core/src/main/config/log4j2.properties#L17C25-L17C25 (eg `\"` everywhere) so the actual print it's not going to be nice looking (like the `request.body` is today)

Reply:
I don't think caching makes any difference in this case.
The caching that relies on the **token** principal only influences which realm is consulted first for that **token** principal.

Given two different JWTs for ultimately the same **user** principal, whether the second JWT ought to first try the realm that authenticated the first one is IMO very debatable: it depends on how different the JWTs are..., as, before the JWTs are authenticated, even minor differences can lead to different **user** principals, in which case the second one should not first try the realm of the first one...

Reply:
> I think should just base64 decode the middle part (the raw claim set) use that as-is. 

I've also pushed https://github.com/elastic/elasticsearch/pull/101333/commits/943113c95ab965b0b8c5d88461a6c266194d3b0e .
One offshoot of it is that the json claim mapping now happens at token authn time rather than token parse time (nothing wrong, just different, and worth to call out).

Reply:
> The caching that relies on the token principal only influences which realm is consulted first for that token principal.

true, but for use cases with high number of JWT realms the signature validation can get quite expensive.  

> it's long and contains date claims such as exp, nbf, iat which are not useful to identifying the principal (in addition to other claims like that, but that we cannot clearly distinguish)

i realize now that including these fields such as exp, nbf, iat can greatly diminish the ability for us to apply the smart realm ordering since workflows that emit one time use JWT (or any short lived token) will have very low cache hits. I think those values have merit to be present the audit log, but only provide minor value.

> eg \" everywhere

ahh... i think you are right. when going JSON->FILE->JSON those will be treated as string literals not as escaped JSON (by default).

For those reasons, I am in favor of [2b47c9d](https://github.com/elastic/elasticsearch/pull/101333/commits/2b47c9d443a7ec5d884beea30a57474f7879eac2). Sorry for the back and forth here, and thanks for pushing for the best answer.

Reply:
Okay, I've reverted https://github.com/elastic/elasticsearch/commit/943113c95ab965b0b8c5d88461a6c266194d3b0e , so [2b47c9d](https://github.com/elastic/elasticsearch/pull/101333/commits/2b47c9d443a7ec5d884beea30a57474f7879eac2) is effective.

On top of that, I've also added extra optimization logic to the `JwtRealm` for the case where a **malformed** JWT is tried to be parsed multiple times by all the JWT realms in the chain (and it fails identically every time), see `JwtRealm#latestMalformedJWTThreadLocal`.
Let me know if you prefer to have that optimization in this PR or not.